File: 2021/2021W16/PreppindataW16.py
import pandas as pd
import re

df = pd.read_csv('F:\\Data\\PL Fixtures.csv').dropna()
big6=['Arsenal','Chelsea','Liverpool','Man Utd','Man City','Spurs']

def get_position(df):
    df['Home Score'] = df['Result'].apply(lambda x: int(re.search('^(\d+)',x).group(1)))
    df['Away Score'] = df['Result'].apply(lambda x: int(re.search('(\d+)$',x).group(1)))
    df['Home Goal Difference'] = df['Home Score'] - df['Away Score']
    df['Away Goal Difference'] = -1*df['Home Goal Difference']
    df['Home Point'] = df['Home Goal Difference'].apply(lambda x: 3 if x>0 else 1 if x==0 else 0)
    df['Away Point'] = df['Away Goal Difference'].apply(lambda x: 3 if x>0 else 1 if x==0 else 0)
    home = df[[c for c in df.columns if c[0:4]=='Home']].copy()
    home.columns = ['Team', 'Score', 'Goal Difference', 'Total Points']
    away = df[[c for c in df.columns if c[0:4]=='Away']].copy()
    away.columns = home.columns 
    
    final = pd.concat([home, away])
    final['Total Games Played'] = 1
    final = final.groupby('Team', as_index=False).agg({'Total Games Played':'sum', 'Total Points':'sum', 'Goal Difference':'sum'})
    final['Position'] = final[['Total Points', 'Goal Difference']].apply(tuple,axis=1).rank(method='dense',ascending=False).astype(int)
    return final[['Position', 'Team', 'Total Games Played', 'Total Points', 'Goal Difference']].copy()

finalA = get_position(df)

finalB = get_position(df[~(df['Home Team'].isin(big6) | df['Away Team'].isin(big6))].copy())
finalB = pd.merge(finalB, finalA[['Team','Position']].rename(columns={'Position':'Position Original'}), on='Team')
finalB['Position Change'] = finalB['Position Original'] - finalB['Position'] 
finalB = finalB[['Position Change', 'Position', 'Team', 'Total Games Played', 'Total Points', 'Goal Difference']]

File: 2021/2021W16/preppindataW16.R
library(readr)
library(dplyr)
library(tidyr)
library(stringr)

data <- read_csv("F:/Data/PL Fixtures.csv") %>% drop_na() 
big6 <- c('Arsenal','Chelsea','Liverpool','Man Utd','Man City','Spurs')
  
get_position <- function(df) {
  df <- df %>%
    mutate('Home Score' = as.integer(str_extract(`Result`, '^(\\d+)')),
           'Away Score' = as.integer(str_extract(`Result`, '(\\d+)$')),
           'Home Goal Difference' = `Home Score` - `Away Score`,
           'Away Goal Difference' = -1*`Home Goal Difference`,
           'Home Point' = if_else(`Home Goal Difference`>0, 3, if_else(`Home Goal Difference`==0, 1, 0)),
           'Away Point' = if_else(`Away Goal Difference`>0, 3, if_else(`Away Goal Difference`==0, 1, 0)))
  
  fin <- bind_rows( df %>% select(starts_with('Home')),
                    df %>% select(starts_with('Away'))) %>%
    mutate_if(is.numeric, list(~replace_na(.,0))) %>%
    mutate('Team' = if_else(is.na(`Away Team`),`Home Team`,`Away Team`)) %>%
    group_by(`Team`) %>%
    summarise('Total Games Played' = n(),
              'Total Points' = sum(`Home Point`+`Away Point`),
              'Goal Difference' = sum(`Home Goal Difference`+`Away Goal Difference`), .groups='drop') %>%
    mutate('Position' = as.integer(rank(interaction(desc(`Total Points`), desc(`Goal Difference`), lex.order=T)))) %>%
    select(c('Position', 'Team', 'Total Games Played', 'Total Points', 'Goal Difference')) 
  return(fin)
}  
  
finA <- get_position(data) %>% 
  arrange(`Position`)

finB <- get_position(data %>% 
                       filter(!(`Home Team` %in% big6 | `Away Team` %in% big6))) %>%
  merge(.,finA %>% select(c('Team', 'Position')) %>% rename('Position Original' = `Position`), by='Team') %>%
  mutate('Position change' = `Position Original` - `Position`) %>%
  select(c('Position change', 'Position', 'Team', 'Total Games Played', 'Total Points', 'Goal Difference')) %>% 
  arrange(`Position`)

View(finA)
View(finB)
